In [1]:
import pandas as pd

In [39]:
import numpy as np

In [2]:
import regex as re

In [3]:
titanic_db = pd.read_csv('../Data/train.csv')


In [4]:
titanic_db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [5]:
titanic_db.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [6]:
change_db=titanic_db.copy(deep=True)


In [7]:
change_db.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [8]:
change_db.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [9]:
pattern = (
    r'^(?P<LastName>[^,]+),\s*'
    r'(?P<Title>\w+)\.?\s*'
    r'(?P<HusbandName>[^(]+)?'
    r'(?:\((?P<ActualName>[^)]+)\))?'
)

extract= change_db['Name'].str.extract(pattern)
extract['HusbandName'] = extract['HusbandName'].str.strip()

change_db2=pd.concat([change_db,extract],axis=1)
change_db2.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked   LastName Title  \
0      0         A/5 21171   7.2500   NaN        S     Braund    Mr   
1      0          PC 17599  71.2833   C85        C    Cumings   Mrs   
2      0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen  Miss   
3      0            113803  53.1000  C123        S   Futrelle   Mrs   
4      0            373450   8.0500   NaN        S      Allen    Mr   

     HusbandName              ActualName  
0    Owen Harris                     NaN  
1   John Bradley  Florence Briggs Thayer  
2          Laina                     NaN  
3  Jacques Heath           Lily May Peel  
4  William Henry                     NaN

In [10]:
change_db2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
 12  LastName     891 non-null    object 
 13  Title        891 non-null    object 
 14  HusbandName  875 non-null    object 
 15  ActualName   143 non-null    object 
dtypes: float64(2), int64(5), object(9)
memory usage: 111.5+ KB


In [11]:
change_db2.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
LastName         0
Title            0
HusbandName     16
ActualName     748
dtype: int64

In [12]:
change_db2.loc[change_db2['HusbandName'].isnull()]

PassengerId  Survived  Pclass  \
15            16         1       2   
66            67         1       2   
166          167         1       1   
190          191         1       2   
259          260         1       2   
272          273         1       2   
327          328         1       2   
362          363         0       3   
367          368         1       3   
483          484         1       3   
516          517         1       2   
556          557         1       1   
564          565         0       3   
772          773         0       2   
823          824         1       3   
865          866         1       2   

                                                  Name     Sex   Age  SibSp  \
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
66                        Nye, Mrs. (Elizabeth Ramell)  female  29.0      0   
166             Chibnall, Mrs. (Edith Martha Bowerman)  female   NaN      0   
190                                Pinsky, Mrs. (Rosa)  female  32.0      0   
259                        Parrish, Mrs. (Lutie Davis)  female  50.0      0   
272          Mellinger, Mrs. (Elizabeth Anne Maidment)  female  41.0      0   
327                            Ball, Mrs. (Ada E Hall)  female  36.0      0   
362                    Barbara, Mrs. (Catherine David)  female  45.0      0   
367                     Moussa, Mrs. (Mantoura Boulos)  female   NaN      0   
483                             Turkula, Mrs. (Hedwig)  female  63.0      0   
516                       Lemore, Mrs. (Amelia Milley)  female  34.0      0   
556  Duff Gordon, Lady. (Lucille Christiana Sutherl...  female  48.0      1   
564                     Meanwell, Miss. (Marion Ogden)  female   NaN      0   
772                                  Mack, Mrs. (Mary)  female  57.0      0   
823                                 Moor, Mrs. (Beila)  female  27.0      0   
865                           Bystrom, Mrs. (Karolina)  female  42.0      0   

     Parch             Ticket     Fare Cabin Embarked     LastName Title  \
15       0             248706  16.0000   NaN        S      Hewlett   Mrs   
66       0         C.A. 29395  10.5000   F33        S          Nye   Mrs   
166      1             113505  55.0000   E33        S     Chibnall   Mrs   
190      0             234604  13.0000   NaN        S       Pinsky   Mrs   
259      1             230433  26.0000   NaN        S      Parrish   Mrs   
272      1             250644  19.5000   NaN        S    Mellinger   Mrs   
327      0              28551  13.0000     D        S         Ball   Mrs   
362      1               2691  14.4542   NaN        C      Barbara   Mrs   
367      0               2626   7.2292   NaN        C       Moussa   Mrs   
483      0               4134   9.5875   NaN        S      Turkula   Mrs   
516      0         C.A. 34260  10.5000   F33        S       Lemore   Mrs   
556      0              11755  39.6000   A16        C  Duff Gordon  Lady   
564      0  SOTON/O.Q. 392087   8.0500   NaN        S     Meanwell  Miss   
772      0        S.O./P.P. 3  10.5000   E77        S         Mack   Mrs   
823      1             392096  12.4750  E121        S         Moor   Mrs   
865      0             236852  13.0000   NaN        S      Bystrom   Mrs   

    HusbandName                     ActualName  
15          NaN                Mary D Kingcome  
66          NaN               Elizabeth Ramell  
166         NaN          Edith Martha Bowerman  
190         NaN                           Rosa  
259         NaN                    Lutie Davis  
272         NaN        Elizabeth Anne Maidment  
327         NaN                     Ada E Hall  
362         NaN                Catherine David  
367         NaN                Mantoura Boulos  
483         NaN                         Hedwig  
516         NaN                  Amelia Milley  
556         NaN  Lucille Christiana Sutherland  
564         NaN                   Marion Ogden  
772         NaN                    

In [13]:
change_db2['Sex'].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

In [14]:
change_db2['Pclass'].value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [15]:
change_db2.loc[change_db2['Survived']==1, 'Sex'].value_counts()

Sex
female    233
male      109
Name: count, dtype: int64

In [16]:
change_db2.loc[((change_db2['Age'].isnull())),'Sex'].value_counts()

Sex
male      124
female     53
Name: count, dtype: int64

In [17]:
change_db2.loc[((change_db2['Age'].isnull())),'Pclass'].value_counts()

Pclass
3    136
1     30
2     11
Name: count, dtype: int64

In [18]:
change_db2.loc[((change_db2['Age'].isnull())),'Title'].value_counts()

Title
Mr        119
Miss       36
Mrs        17
Master      4
Dr          1
Name: count, dtype: int64

In [19]:
change_db2.loc[((change_db2['Age'].isnull())&(change_db2['Title']=='Dr'))]

PassengerId  Survived  Pclass                       Name   Sex  Age  \
766          767         0       1  Brewe, Dr. Arthur Jackson  male  NaN   

     SibSp  Parch  Ticket  Fare Cabin Embarked LastName Title     HusbandName  \
766      0      0  112379  39.6   NaN        C    Brewe    Dr  Arthur Jackson   

    ActualName  
766        NaN

In [20]:
change_db2.loc[((change_db2['Title']=='Dr'))]

PassengerId  Survived  Pclass                           Name     Sex  \
245          246         0       1    Minahan, Dr. William Edward    male   
317          318         0       2           Moraweck, Dr. Ernest    male   
398          399         0       2               Pain, Dr. Alfred    male   
632          633         1       1      Stahelin-Maeglin, Dr. Max    male   
660          661         1       1  Frauenthal, Dr. Henry William    male   
766          767         0       1      Brewe, Dr. Arthur Jackson    male   
796          797         1       1    Leader, Dr. Alice (Farnham)  female   

      Age  SibSp  Parch    Ticket      Fare Cabin Embarked          LastName  \
245  44.0      2      0     19928   90.0000   C78        Q           Minahan   
317  54.0      0      0     29011   14.0000   NaN        S          Moraweck   
398  23.0      0      0    244278   10.5000   NaN        S              Pain   
632  32.0      0      0     13214   30.5000   B50        C  Stahelin-Maeglin   
660  50.0      2      0  PC 17611  133.6500   NaN        S        Frauenthal   
766   NaN      0      0    112379   39.6000   NaN        C             Brewe   
796  49.0      0      0     17465   25.9292   D17        S            Leader   

    Title     HusbandName ActualName  
245    Dr  William Edward        NaN  
317    Dr          Ernest        NaN  
398    Dr          Alfred        NaN  
632    Dr             Max        NaN  
660    Dr   Henry William        NaN  
766    Dr  Arthur Jackson        NaN  
796    Dr           Alice    Farnham

There are 7 `Dr.` passengers, where **1** is *female* and 6 are *male* Dr. passengers. 


In [21]:
change_db2.loc[((change_db2['Sex']=='male')&(change_db2['Title']=='Dr'))].describe()

PassengerId  Survived    Pclass        Age     SibSp  Parch        Fare
count     6.000000  6.000000  6.000000   5.000000  6.000000    6.0    6.000000
mean    504.000000  0.333333  1.333333  40.600000  0.666667    0.0   53.041667
std     211.018483  0.516398  0.516398  12.876335  1.032796    0.0   48.757506
min     246.000000  0.000000  1.000000  23.000000  0.000000    0.0   10.500000
25%     338.250000  0.000000  1.000000  32.000000  0.000000    0.0   18.125000
50%     516.000000  0.000000  1.000000  44.000000  0.000000    0.0   35.050000
75%     654.000000  0.750000  1.750000  50.000000  1.500000    0.0   77.400000
max     767.000000  1.000000  2.000000  54.000000  2.000000    0.0  133.650000

In [22]:
change_db_cleaned=change_db2.copy(deep=True)

For final cleaning created a table `change_db_cleaned` by creating a copy of `change_db2`

In [23]:
change_db_cleaned.loc[((change_db2['Age'].isnull())&(change_db2['Title']=='Dr')),'Age']=41

In [24]:
change_db_cleaned.loc[((change_db2['Sex']=='male')&(change_db2['Title']=='Dr'))].describe()

PassengerId  Survived    Pclass        Age     SibSp  Parch        Fare
count     6.000000  6.000000  6.000000   6.000000  6.000000    6.0    6.000000
mean    504.000000  0.333333  1.333333  40.666667  0.666667    0.0   53.041667
std     211.018483  0.516398  0.516398  11.518102  1.032796    0.0   48.757506
min     246.000000  0.000000  1.000000  23.000000  0.000000    0.0   10.500000
25%     338.250000  0.000000  1.000000  34.250000  0.000000    0.0   18.125000
50%     516.000000  0.000000  1.000000  42.500000  0.000000    0.0   35.050000
75%     654.000000  0.750000  1.750000  48.500000  1.500000    0.0   77.400000
max     767.000000  1.000000  2.000000  54.000000  2.000000    0.0  133.650000

`Dr.` title (1 null): Impute with **40.6** rounding it to **41**— mean age of male Dr. passengers.

In [25]:
change_db2.loc[((change_db2['Age'].isnull())&(change_db2['Title']=='Master'))]

PassengerId  Survived  Pclass  \
65            66         1       3   
159          160         0       3   
176          177         0       3   
709          710         1       3   

                                                  Name   Sex  Age  SibSp  \
65                            Moubarek, Master. Gerios  male  NaN      1   
159                         Sage, Master. Thomas Henry  male  NaN      8   
176                      Lefebre, Master. Henry Forbes  male  NaN      3   
709  Moubarek, Master. Halim Gonios ("William George")  male  NaN      1   

     Parch    Ticket     Fare Cabin Embarked  LastName   Title   HusbandName  \
65       1      2661  15.2458   NaN        C  Moubarek  Master        Gerios   
159      2  CA. 2343  69.5500   NaN        S      Sage  Master  Thomas Henry   
176      1      4133  25.4667   NaN        S   Lefebre  Master  Henry Forbes   
709      1      2661  15.2458   NaN        C  Moubarek  Master  Halim Gonios   

           ActualName  
65                NaN  
159               NaN  
176               NaN  
709  "William George"

`Master` title of 3rd class has all 4 null values of age

In [26]:
change_db2.loc[((change_db2['Title']=='Master')&(change_db2['Pclass']==3))].describe()

PassengerId   Survived  Pclass        Age      SibSp      Parch  \
count    28.000000  28.000000    28.0  24.000000  28.000000  28.000000   
mean    388.285714   0.392857     3.0   5.350833   2.928571   1.357143   
std     314.309927   0.497347     0.0   3.593608   1.942289   0.558721   
min       8.000000   0.000000     3.0   0.420000   0.000000   0.000000   
25%     151.500000   0.000000     3.0   2.000000   1.000000   1.000000   
50%     270.500000   0.000000     3.0   4.000000   3.500000   1.000000   
75%     761.000000   1.000000     3.0   9.000000   4.000000   2.000000   
max     870.000000   1.000000     3.0  12.000000   8.000000   2.000000   

            Fare  
count  28.000000  
mean   28.177232  
std    13.898033  
min     8.516700  
25%    15.900000  
50%    28.512500  
75%    33.462500  
max    69.550000

In [27]:
change_db_cleaned.loc[(change_db2['Title']=='Master')&(change_db2['Age'].isnull()),'Age']=5

In [28]:
change_db_cleaned.loc[((change_db2['Title']=='Master')&(change_db2['Pclass']==3))].describe()

PassengerId   Survived  Pclass        Age      SibSp      Parch  \
count    28.000000  28.000000    28.0  28.000000  28.000000  28.000000   
mean    388.285714   0.392857     3.0   5.300714   2.928571   1.357143   
std     314.309927   0.497347     0.0   3.319105   1.942289   0.558721   
min       8.000000   0.000000     3.0   0.420000   0.000000   0.000000   
25%     151.500000   0.000000     3.0   2.750000   1.000000   1.000000   
50%     270.500000   0.000000     3.0   5.000000   3.500000   1.000000   
75%     761.000000   1.000000     3.0   8.250000   4.000000   2.000000   
max     870.000000   1.000000     3.0  12.000000   8.000000   2.000000   

            Fare  
count  28.000000  
mean   28.177232  
std    13.898033  
min     8.516700  
25%    15.900000  
50%    28.512500  
75%    33.462500  
max    69.550000

`Master` in 3rd class (4 nulls): Impute with **5.3** round off to **5** — mean age of Master passengers in 3rd class.


In [ ]:
change_db2.loc[(change_db2['Title']=='Mrs')]

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
8              9         1       3   
9             10         1       2   
15            16         1       2   
..           ...       ...     ...   
871          872         1       1   
874          875         1       2   
879          880         1       1   
880          881         1       2   
885          886         0       3   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
874              Abelson, Mrs. Samuel (Hannah Wizosky)  female  28.0      1   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   

     Parch     Ticket     Fare Cabin Embarked  LastName Title  \
1        0   PC 17599  71.2833   C85        C   Cumings   Mrs   
3        0     113803  53.1000  C123        S  Futrelle   Mrs   
8        2     347742  11.1333   NaN        S   Johnson   Mrs   
9        0     237736  30.0708   NaN        C    Nasser   Mrs   
15       0     248706  16.0000   NaN        S   Hewlett   Mrs   
..     ...        ...      ...   ...      ...       ...   ...   
871      1      11751  52.5542   D35        S  Beckwith   Mrs   
874      0  P/PP 3381  24.0000   NaN        C   Abelson   Mrs   
879      1      11767  83.1583   C50        C    Potter   Mrs   
880      1     230433  26.0000   NaN        S   Shelley   Mrs   
885      5     382652  29.1250   NaN        Q      Rice   Mrs   

         HusbandName                 ActualName  
1       John Bradley     Florence Briggs Thayer  
3      Jacques Heath              Lily May Peel  
8            Oscar W  Elisabeth Vilhelmina Berg  
9           Nicholas                Adele Achem  
15               NaN            Mary D Kingcome  
..               ...                        ...  
871  Richard Leonard            Sallie Monypeny  
874           Samuel             Hannah Wizosky  
879        Thomas Jr       Lily Alexenia Wilson  
880          William       Imanita Parrish Hall  
885          William            Margaret Norton  

[125 rows x 16 columns]

In [73]:
change_db2.loc[(change_db2['Title']=='Mrs'),'Age'].describe()

count    108.000000
mean      35.898148
std       11.433628
min       14.000000
25%       27.750000
50%       35.000000
75%       44.000000
max       63.000000
Name: Age, dtype: float64

In [74]:
change_db2.loc[(change_db2['Title']=='Mrs'),'Age'].median()

np.float64(35.0)

In [83]:
change_db2.loc[((change_db2['Age'].notnull())&(change_db2['Title']=='Mrs')),'Pclass'].value_counts()

Pclass
2    41
1    34
3    33
Name: count, dtype: int64

In [82]:
change_db2.loc[((change_db2['Age'].isnull())&(change_db2['Title']=='Mrs')),'Pclass'].value_counts()

Pclass
3    9
1    8
Name: count, dtype: int64

In [87]:
change_db2.loc[((change_db2['Pclass']==1)&(change_db2['Title']=='Mrs')),'Age'].describe()

count    34.000000
mean     40.882353
std      12.643754
min      17.000000
25%      35.000000
50%      41.500000
75%      50.750000
max      62.000000
Name: Age, dtype: float64

In [88]:
change_db2.loc[((change_db2['Pclass']==1)&(change_db2['Title']=='Mrs')),'Age'].median()

np.float64(41.5)

In [124]:
change_db2.loc[(change_db2['Pclass']==1)&(change_db2['Title']=='Mrs'),['Pclass','Name','Age','SibSp','Parch','Fare']].sort_values('SibSp')

Pclass                                               Name   Age  SibSp  \
194       1          Brown, Mrs. James Joseph (Margaret Tobin)  44.0      0   
166       1             Chibnall, Mrs. (Edith Martha Bowerman)   NaN      0   
268       1      Graham, Mrs. William Thompson (Edith Junkins)  58.0      0   
256       1                     Thorne, Mrs. Gertrude Maybelle   NaN      0   
299       1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)  50.0      0   
523       1    Hippach, Mrs. Louis Albert (Ida Sophia Fischer)  44.0      0   
829       1          Stone, Mrs. George Nelson (Martha Evelyn)  62.0      0   
779       1  Robert, Mrs. Edward Scott (Elisabeth Walton Mc...  43.0      0   
879       1      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  56.0      0   
862       1  Swift, Mrs. Frederick Joel (Margaret Welles Ba...  48.0      0   
31        1     Spencer, Mrs. William Augustus (Marie Eugenie)   NaN      1   
291       1            Bishop, Mrs. Dickinson H (Helen Walton)  19.0      1   
230       1       Harris, Mrs. Henry Birkhardt (Irene Wallach)  35.0      1   
151       1                  Pears, Mrs. Thomas (Edith Wearne)  22.0      1   
1         1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0      1   
52        1           Harper, Mrs. Henry Sleeper (Myna Haxtun)  49.0      1   
366       1   Warren, Mrs. Frank Manley (Anna Sophia Atkinson)  60.0      1   
375       1              Meyer, Mrs. Edgar Joseph (Leila Saks)   NaN      1   
457       1                  Kenyon, Mrs. Frederick R (Marion)   NaN      1   
383       1  Holverson, Mrs. Alexander Oskar (Mary Aline To...  35.0      1   
486       1    Hoyt, Mrs. Frederick Maxfield (Jane Anne Forby)  35.0      1   
498       1    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  25.0      1   
513       1     Rothschild, Mrs. Martin (Elizabeth L. Barrett)  54.0      1   
558       1             Taussig, Mrs. Emil (Tillie Mandelbaum)  39.0      1   
577       1          Silvey, Mrs. William Baird (Alice Munger)  39.0      1   
581       1  Thayer, Mrs. John Borland (Marian Longstreth M...  39.0      1   
591       1    Stephenson, Mrs. Walter Bertram (Martha Eustis)  52.0      1   
669       1  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)   NaN      1   
307       1  Penasco y Castellana, Mrs. Victor de Satode (M...  17.0      1   
319       1  Spedden, Mrs. Frederic Oakley (Margaretta Corn...  40.0      1   
334       1  Frauenthal, Mrs. Henry William (Clara Heinshei...   NaN      1   
3         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0      1   
856       1         Wick, Mrs. George Dennick (Mary Hitchcock)  45.0      1   
765       1               Hogeboom, Mrs. John C (Anna Andrews)  51.0      1   
700       1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  18.0      1   
763       1          Carter, Mrs. William Ernest (Lucile Polk)  36.0      1   
820       1  Hays, Mrs. Charles Melville (Clara Jennings Gr...  52.0      1   
809       1     Chambers, Mrs. Norman Campbell (Bertha Griggs)  33.0      1   
849       1       Goldenberg, Mrs. Samuel L (Edwiga Grabowska)   NaN      1   
781       1          Dick, Mrs. Albert Adrian (Vera Gillespie)  17.0      1   
871       1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  47.0      1   
571       1      Appleton, Mrs. Edward Dale (Charlotte Lamson)  53.0      2   

     Parch      Fare  
194      0   27.7208  
166      1   55.0000  
268      1  153.4625  
256      0   79.2000  
299      1  247.5208  
523      1   57.9792  
829      0   80.0000  
779      1  211.3375  
879      1   83.1583  
862      0   25.9292  
31       0  146.5208  
291      0   91.0792  
230      0   83.4750  
151      0   66.6000  
1        0   71.2833  
52       0   76.7292  
366      0   75.2500  
375      0   82.1708  
457      0   51.8625  
383      0   52.0000  
486      0   90.0000  
498      2  151.5500  
513      0   59.4000  
558      1   79.6500  
577      0   55.9000  
581      

In [125]:
change_db2.loc[(change_db2['Pclass']==1)&(change_db2['Title']=='Mrs')&(change_db2['SibSp']==0),['Pclass','Name','Age','SibSp','Parch','Fare']]

Pclass                                               Name   Age  SibSp  \
166       1             Chibnall, Mrs. (Edith Martha Bowerman)   NaN      0   
194       1          Brown, Mrs. James Joseph (Margaret Tobin)  44.0      0   
256       1                     Thorne, Mrs. Gertrude Maybelle   NaN      0   
268       1      Graham, Mrs. William Thompson (Edith Junkins)  58.0      0   
299       1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)  50.0      0   
523       1    Hippach, Mrs. Louis Albert (Ida Sophia Fischer)  44.0      0   
779       1  Robert, Mrs. Edward Scott (Elisabeth Walton Mc...  43.0      0   
829       1          Stone, Mrs. George Nelson (Martha Evelyn)  62.0      0   
862       1  Swift, Mrs. Frederick Joel (Margaret Welles Ba...  48.0      0   
879       1      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  56.0      0   

     Parch      Fare  
166      1   55.0000  
194      0   27.7208  
256      0   79.2000  
268      1  153.4625  
299      1  247.5208  
523      1   57.9792  
779      1  211.3375  
829      0   80.0000  
862      0   25.9292  
879      1   83.1583

In [138]:
change_db2.loc[(change_db2['Pclass']==1)&(change_db2['Title']=='Mrs')&(change_db2['SibSp']==0)&(change_db2['Parch']==0),['Pclass','Name','Age','SibSp','Parch','Fare']]

Pclass                                               Name   Age  SibSp  \
194       1          Brown, Mrs. James Joseph (Margaret Tobin)  44.0      0   
256       1                     Thorne, Mrs. Gertrude Maybelle   NaN      0   
829       1          Stone, Mrs. George Nelson (Martha Evelyn)  62.0      0   
862       1  Swift, Mrs. Frederick Joel (Margaret Welles Ba...  48.0      0   

     Parch     Fare  
194      0  27.7208  
256      0  79.2000  
829      0  80.0000  
862      0  25.9292

In [117]:
change_db2.loc[(change_db2['Pclass']==1)&(change_db2['Title']=='Mrs')&(change_db2['SibSp']==0)&(change_db2['Parch']==0),['Pclass','Name','Age','SibSp','Parch','Fare']].describe()

Pclass        Age  SibSp  Parch       Fare
count     4.0   3.000000    4.0    4.0   4.000000
mean      1.0  51.333333    0.0    0.0  53.212500
std       0.0   9.451631    0.0    0.0  30.480188
min       1.0  44.000000    0.0    0.0  25.929200
25%       1.0  46.000000    0.0    0.0  27.272900
50%       1.0  48.000000    0.0    0.0  53.460400
75%       1.0  55.000000    0.0    0.0  79.400000
max       1.0  62.000000    0.0    0.0  80.000000

In [133]:
change_db_cleaned.loc[(change_db_cleaned['Pclass']==1)&(change_db_cleaned['Title']=='Mrs')&(change_db_cleaned['SibSp']==0)&(change_db_cleaned['Parch']==0)&(change_db_cleaned['Age'].isnull()),['Age']]=57

In [135]:
change_db_cleaned.loc[(change_db_cleaned['Pclass']==1)&(change_db_cleaned['Title']=='Mrs')&(change_db_cleaned['SibSp']==0)&(change_db_cleaned['Parch']==0),['Age']].describe()

Age
count   4.000000
mean   52.750000
std     8.220908
min    44.000000
25%    47.000000
50%    52.500000
75%    58.250000
max    62.000000

`Mrs` title in `1st class` and has 0 -`SibSp` and 0-`Parch` **(1 null)**: Impute with **57**- middle value between mean age of group and max age of group. 

In [139]:
change_db2.loc[(change_db2['Pclass']==1)&(change_db2['Title']=='Mrs')&(change_db2['SibSp']==0)&(change_db2['Parch']==1),['Pclass','Name','Age','SibSp','Parch','Fare','Cabin']]

Pclass                                               Name   Age  SibSp  \
166       1             Chibnall, Mrs. (Edith Martha Bowerman)   NaN      0   
268       1      Graham, Mrs. William Thompson (Edith Junkins)  58.0      0   
299       1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)  50.0      0   
523       1    Hippach, Mrs. Louis Albert (Ida Sophia Fischer)  44.0      0   
779       1  Robert, Mrs. Edward Scott (Elisabeth Walton Mc...  43.0      0   
879       1      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  56.0      0   

     Parch      Fare    Cabin  
166      1   55.0000      E33  
268      1  153.4625     C125  
299      1  247.5208  B58 B60  
523      1   57.9792      B18  
779      1  211.3375       B3  
879      1   83.1583      C50

In [136]:
change_db2[change_db2['Cabin']=='E33']

PassengerId  Survived  Pclass                                    Name  \
166          167         1       1  Chibnall, Mrs. (Edith Martha Bowerman)   
356          357         1       1             Bowerman, Miss. Elsie Edith   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  LastName Title  \
166  female   NaN      0      1  113505  55.0   E33        S  Chibnall   Mrs   
356  female  22.0      0      1  113505  55.0   E33        S  Bowerman  Miss   

     HusbandName             ActualName  
166          NaN  Edith Martha Bowerman  
356  Elsie Edith                    NaN

In [147]:
change_db2[change_db2['Cabin']=='B58 B60']

PassengerId  Survived  Pclass  \
118          119         0       1   
299          300         1       1   

                                                Name     Sex   Age  SibSp  \
118                         Baxter, Mr. Quigg Edmond    male  24.0      0   
299  Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0      0   

     Parch    Ticket      Fare    Cabin Embarked LastName Title   HusbandName  \
118      1  PC 17558  247.5208  B58 B60        C   Baxter    Mr  Quigg Edmond   
299      1  PC 17558  247.5208  B58 B60        C   Baxter   Mrs         James   

                     ActualName  
118                         NaN  
299  Helene DeLaudeniere Chaput

In [127]:
change_db2.loc[(change_db2['Pclass']==1)&(change_db2['Title']=='Mrs')&(change_db2['SibSp']==0)&(change_db2['Parch']==1),['Pclass','Name','Age','SibSp','Parch','Fare']].describe()

Pclass        Age  SibSp  Parch        Fare
count     6.0   5.000000    6.0    6.0    6.000000
mean      1.0  50.200000    0.0    1.0  134.743050
std       0.0   6.797058    0.0    0.0   82.278281
min       1.0  43.000000    0.0    1.0   55.000000
25%       1.0  44.000000    0.0    1.0   64.273975
50%       1.0  50.000000    0.0    1.0  118.310400
75%       1.0  56.000000    0.0    1.0  196.868750
max       1.0  58.000000    0.0    1.0  247.520800

In [149]:
change_db_cleaned.loc[(change_db_cleaned['Pclass']==1)&(change_db_cleaned['Title']=='Mrs')&(change_db_cleaned['SibSp']==0)&(change_db_cleaned['Parch']==1)&(change_db_cleaned['Age'].isnull()),['Age']]=50

In [150]:
change_db_cleaned.loc[(change_db_cleaned['Pclass']==1)&(change_db_cleaned['Title']=='Mrs')&(change_db_cleaned['SibSp']==0)&(change_db_cleaned['Parch']==1),['Age']].describe()

Age
count   6.000000
mean   50.166667
std     6.080022
min    43.000000
25%    45.500000
50%    50.000000
75%    54.500000
max    58.000000

`Mrs` title in `1st class` and has 0 -`SibSp` and 1-`Parch` **(1 null)**: Impute with **50.2** round off to 50- mean age of group and comparing with a person from this group who is having a son of same age group as her daughters' age. 

In [152]:
change_db2.loc[(change_db2['Pclass']==1)&(change_db2['Title']=='Mrs')&(change_db2['SibSp']==1),['Pclass','Name','Age','SibSp','Parch','Fare']].sort_values('Parch')

Pclass                                               Name   Age  SibSp  \
1         1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0      1   
3         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0      1   
31        1     Spencer, Mrs. William Augustus (Marie Eugenie)   NaN      1   
52        1           Harper, Mrs. Henry Sleeper (Myna Haxtun)  49.0      1   
151       1                  Pears, Mrs. Thomas (Edith Wearne)  22.0      1   
230       1       Harris, Mrs. Henry Birkhardt (Irene Wallach)  35.0      1   
291       1            Bishop, Mrs. Dickinson H (Helen Walton)  19.0      1   
307       1  Penasco y Castellana, Mrs. Victor de Satode (M...  17.0      1   
334       1  Frauenthal, Mrs. Henry William (Clara Heinshei...   NaN      1   
366       1   Warren, Mrs. Frank Manley (Anna Sophia Atkinson)  60.0      1   
375       1              Meyer, Mrs. Edgar Joseph (Leila Saks)   NaN      1   
383       1  Holverson, Mrs. Alexander Oskar (Mary Aline To...  35.0      1   
486       1    Hoyt, Mrs. Frederick Maxfield (Jane Anne Forby)  35.0      1   
457       1                  Kenyon, Mrs. Frederick R (Marion)   NaN      1   
577       1          Silvey, Mrs. William Baird (Alice Munger)  39.0      1   
513       1     Rothschild, Mrs. Martin (Elizabeth L. Barrett)  54.0      1   
849       1       Goldenberg, Mrs. Samuel L (Edwiga Grabowska)   NaN      1   
809       1     Chambers, Mrs. Norman Campbell (Bertha Griggs)  33.0      1   
669       1  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)   NaN      1   
591       1    Stephenson, Mrs. Walter Bertram (Martha Eustis)  52.0      1   
700       1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  18.0      1   
765       1               Hogeboom, Mrs. John C (Anna Andrews)  51.0      1   
781       1          Dick, Mrs. Albert Adrian (Vera Gillespie)  17.0      1   
319       1  Spedden, Mrs. Frederic Oakley (Margaretta Corn...  40.0      1   
871       1   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  47.0      1   
820       1  Hays, Mrs. Charles Melville (Clara Jennings Gr...  52.0      1   
558       1             Taussig, Mrs. Emil (Tillie Mandelbaum)  39.0      1   
581       1  Thayer, Mrs. John Borland (Marian Longstreth M...  39.0      1   
856       1         Wick, Mrs. George Dennick (Mary Hitchcock)  45.0      1   
498       1    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  25.0      1   
763       1          Carter, Mrs. William Ernest (Lucile Polk)  36.0      1   

     Parch      Fare  
1        0   71.2833  
3        0   53.1000  
31       0  146.5208  
52       0   76.7292  
151      0   66.6000  
230      0   83.4750  
291      0   91.0792  
307      0  108.9000  
334      0  133.6500  
366      0   75.2500  
375      0   82.1708  
383      0   52.0000  
486      0   90.0000  
457      0   51.8625  
577      0   55.9000  
513      0   59.4000  
849      0   89.1042  
809      0   53.1000  
669      0   52.0000  
591      0   78.2667  
700      0  227.5250  
765      0   77.9583  
781      0   57.0000  
319      1  134.5000  
871      1   52.5542  
820      1   93.5000  
558      1   79.6500  
581      1  110.8833  
856      1  164.8667  
498      2  151.5500  
763      2  120.0000

In [156]:
change_db2.loc[(change_db2['Pclass']==1)&(change_db2['Title']=='Mrs')&(change_db2['SibSp']==1)&(change_db2['Parch']==0)].sort_values('Age')

PassengerId  Survived  Pclass  \
307          308         1       1   
781          782         1       1   
700          701         1       1   
291          292         1       1   
151          152         1       1   
809          810         1       1   
230          231         1       1   
383          384         1       1   
486          487         1       1   
3              4         1       1   
1              2         1       1   
577          578         1       1   
52            53         1       1   
765          766         1       1   
591          592         1       1   
513          514         1       1   
366          367         1       1   
31            32         1       1   
334          335         1       1   
375          376         1       1   
457          458         1       1   
669          670         1       1   
849          850         1       1   

                                                  Name     Sex   Age  SibSp  \
307  Penasco y Castellana, Mrs. Victor de Satode (M...  female  17.0      1   
781          Dick, Mrs. Albert Adrian (Vera Gillespie)  female  17.0      1   
700  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  female  18.0      1   
291            Bishop, Mrs. Dickinson H (Helen Walton)  female  19.0      1   
151                  Pears, Mrs. Thomas (Edith Wearne)  female  22.0      1   
809     Chambers, Mrs. Norman Campbell (Bertha Griggs)  female  33.0      1   
230       Harris, Mrs. Henry Birkhardt (Irene Wallach)  female  35.0      1   
383  Holverson, Mrs. Alexander Oskar (Mary Aline To...  female  35.0      1   
486    Hoyt, Mrs. Frederick Maxfield (Jane Anne Forby)  female  35.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
577          Silvey, Mrs. William Baird (Alice Munger)  female  39.0      1   
52            Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.0      1   
765               Hogeboom, Mrs. John C (Anna Andrews)  female  51.0      1   
591    Stephenson, Mrs. Walter Bertram (Martha Eustis)  female  52.0      1   
513     Rothschild, Mrs. Martin (Elizabeth L. Barrett)  female  54.0      1   
366   Warren, Mrs. Frank Manley (Anna Sophia Atkinson)  female  60.0      1   
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female   NaN      1   
334  Frauenthal, Mrs. Henry William (Clara Heinshei...  female   NaN      1   
375              Meyer, Mrs. Edgar Joseph (Leila Saks)  female   NaN      1   
457                  Kenyon, Mrs. Frederick R (Marion)  female   NaN      1   
669  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)  female   NaN      1   
849       Goldenberg, Mrs. Samuel L (Edwiga Grabowska)  female   NaN      1   

     Parch    Ticket      Fare    Cabin Embarked              LastName Title  \
307      0  PC 17758  108.9000      C65        C  Penasco y Castellana   Mrs   
781      0     17474   57.0000      B20        S                  Dick   Mrs   
700      0  PC 17757  227.5250  C62 C64        C                 Astor   Mrs   
291      0     11967   91.0792      B49        C                Bishop   Mrs   
151      0    113776   66.6000       C2        S                 Pears   Mrs   
809      0    113806   53.1000       E8        S              Chambers   Mrs   
230      0     36973   83.4750      C83        S                Harris   Mrs   
383      0    113789   52.0000      NaN        S             Holverson   Mrs   
486      0     19943   90.0000      C93        S                  Hoyt   Mrs   
3        0    113803   53.1000     C123        S              Futrelle   Mrs   
1        0  PC 17599   71.2833      C85        C               Cumings   Mrs   
577      0     13507   55.9000      E44        S                Silvey   Mrs   
52       0  PC 17572   76.7292      D33        C                Harper   Mrs   
765      0     13502   77.9583      D11        S              Hogeboom   Mr

In [171]:
change_db2.loc[(change_db2['Pclass']==1)&(change_db2['Title']=='Mrs')&(change_db2['SibSp']==1)&(change_db2['Parch']==0),['Pclass','Name','Age','SibSp','Parch','Fare']].describe()

Pclass        Age  SibSp  Parch        Fare
count    23.0  17.000000   23.0   23.0   23.000000
mean      1.0  35.823529    1.0    0.0   84.038043
std       0.0  13.933930    0.0    0.0   40.162538
min       1.0  17.000000    1.0    0.0   51.862500
25%       1.0  22.000000    1.0    0.0   56.450000
50%       1.0  35.000000    1.0    0.0   76.729200
75%       1.0  49.000000    1.0    0.0   89.552100
max       1.0  60.000000    1.0    0.0  227.525000

In [161]:
array=change_db2.loc[(change_db2['Pclass']==1)&(change_db2['Title']=='Mrs')&(change_db2['SibSp']==1)&(change_db2['Parch']==0),'HusbandName'].unique()

In [177]:
change_db2.loc[(change_db2['HusbandName'].isin(array))&(change_db2['Pclass']==1),['Pclass','Name','Age','Title', 'SibSp','Parch','Fare','HusbandName']].sort_values(['HusbandName','Age'])

Pclass                                               Name   Age Title  \
781       1          Dick, Mrs. Albert Adrian (Vera Gillespie)  17.0   Mrs   
690       1                            Dick, Mr. Albert Adrian  31.0    Mr   
383       1  Holverson, Mrs. Alexander Oskar (Mary Aline To...  35.0   Mrs   
35        1                     Holverson, Mr. Alexander Oskar  42.0    Mr   
291       1            Bishop, Mrs. Dickinson H (Helen Walton)  19.0   Mrs   
484       1                            Bishop, Mr. Dickinson H  25.0    Mr   
34        1                            Meyer, Mr. Edgar Joseph  28.0    Mr   
375       1              Meyer, Mrs. Edgar Joseph (Leila Saks)   NaN   Mrs   
712       1                           Taylor, Mr. Elmer Zebley  48.0    Mr   
669       1  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)   NaN   Mrs   
366       1   Warren, Mrs. Frank Manley (Anna Sophia Atkinson)  60.0   Mrs   
486       1    Hoyt, Mrs. Frederick Maxfield (Jane Anne Forby)  35.0   Mrs   
224       1                       Hoyt, Mr. Frederick Maxfield  38.0    Mr   
457       1                  Kenyon, Mrs. Frederick R (Marion)   NaN   Mrs   
230       1       Harris, Mrs. Henry Birkhardt (Irene Wallach)  35.0   Mrs   
62        1                        Harris, Mr. Henry Birkhardt  45.0    Mr   
645       1                          Harper, Mr. Henry Sleeper  48.0    Mr   
52        1           Harper, Mrs. Henry Sleeper (Myna Haxtun)  49.0   Mrs   
660       1                      Frauenthal, Dr. Henry William  50.0    Dr   
334       1  Frauenthal, Mrs. Henry William (Clara Heinshei...   NaN   Mrs   
3         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0   Mrs   
137       1                        Futrelle, Mr. Jacques Heath  37.0    Mr   
1         1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0   Mrs   
765       1               Hogeboom, Mrs. John C (Anna Andrews)  51.0   Mrs   
700       1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  18.0   Mrs   
513       1     Rothschild, Mrs. Martin (Elizabeth L. Barrett)  54.0   Mrs   
724       1                      Chambers, Mr. Norman Campbell  27.0    Mr   
809       1     Chambers, Mrs. Norman Campbell (Bertha Griggs)  33.0   Mrs   
453       1                           Goldenberg, Mr. Samuel L  49.0    Mr   
849       1       Goldenberg, Mrs. Samuel L (Edwiga Grabowska)   NaN   Mrs   
151       1                  Pears, Mrs. Thomas (Edith Wearne)  22.0   Mrs   
307       1  Penasco y Castellana, Mrs. Victor de Satode (M...  17.0   Mrs   
505       1         Penasco y Castellana, Mr. Victor de Satode  18.0    Mr   
591       1    Stephenson, Mrs. Walter Bertram (Martha Eustis)  52.0   Mrs   
31        1     Spencer, Mrs. William Augustus (Marie Eugenie)   NaN   Mrs   
577       1          Silvey, Mrs. William Baird (Alice Munger)  39.0   Mrs   
434       1                          Silvey, Mr. William Baird  50.0    Mr   

     SibSp  Parch      Fare         HusbandName  
781      1      0   57.0000       Albert Adrian  
690      1      0   57.0000       Albert Adrian  
383      1      0   52.0000     Alexander Oskar  
35       1      0   52.0000     Alexander Oskar  
291      1      0   91.0792         Dickinson H  
484      1      0   91.0792         Dickinson H  
34       1      0   82.1708        Edgar Joseph  
375      1      0   82.1708        Edgar Joseph  
712      1      0   52.0000        Elmer Zebley  
669      1      0   52.0000        Elmer Zebley  
366      1      0   75.2500        Frank Manley  
486      1      0   90.0000  Frederick Maxfield  
224      1      0   90.0000  Frederick Maxfield  
457      1      0   51.8625         Frederick R  
230      1      0   83.4750     Henry Birkhardt  
62       1      0   83.4750     Henry Birkhardt  
645      1      0   76.7292       Henry Sleeper  
52       1      0   76.7292       Henry Sleeper  
660      2      0  133.6500       Henry William  
334      1      0  133.6500       Henry 

In [178]:
change_db2.loc[(change_db2['Pclass']==1)&(change_db2['Title']=='Mrs')&(change_db2['SibSp']==1)&(change_db2['Parch']==0),['Pclass','Name','Age','SibSp','Parch','Fare','Cabin']].sort_values('Age')

Pclass                                               Name   Age  SibSp  \
307       1  Penasco y Castellana, Mrs. Victor de Satode (M...  17.0      1   
781       1          Dick, Mrs. Albert Adrian (Vera Gillespie)  17.0      1   
700       1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  18.0      1   
291       1            Bishop, Mrs. Dickinson H (Helen Walton)  19.0      1   
151       1                  Pears, Mrs. Thomas (Edith Wearne)  22.0      1   
809       1     Chambers, Mrs. Norman Campbell (Bertha Griggs)  33.0      1   
230       1       Harris, Mrs. Henry Birkhardt (Irene Wallach)  35.0      1   
383       1  Holverson, Mrs. Alexander Oskar (Mary Aline To...  35.0      1   
486       1    Hoyt, Mrs. Frederick Maxfield (Jane Anne Forby)  35.0      1   
3         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0      1   
1         1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0      1   
577       1          Silvey, Mrs. William Baird (Alice Munger)  39.0      1   
52        1           Harper, Mrs. Henry Sleeper (Myna Haxtun)  49.0      1   
765       1               Hogeboom, Mrs. John C (Anna Andrews)  51.0      1   
591       1    Stephenson, Mrs. Walter Bertram (Martha Eustis)  52.0      1   
513       1     Rothschild, Mrs. Martin (Elizabeth L. Barrett)  54.0      1   
366       1   Warren, Mrs. Frank Manley (Anna Sophia Atkinson)  60.0      1   
31        1     Spencer, Mrs. William Augustus (Marie Eugenie)   NaN      1   
334       1  Frauenthal, Mrs. Henry William (Clara Heinshei...   NaN      1   
375       1              Meyer, Mrs. Edgar Joseph (Leila Saks)   NaN      1   
457       1                  Kenyon, Mrs. Frederick R (Marion)   NaN      1   
669       1  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)   NaN      1   
849       1       Goldenberg, Mrs. Samuel L (Edwiga Grabowska)   NaN      1   

     Parch      Fare    Cabin  
307      0  108.9000      C65  
781      0   57.0000      B20  
700      0  227.5250  C62 C64  
291      0   91.0792      B49  
151      0   66.6000       C2  
809      0   53.1000       E8  
230      0   83.4750      C83  
383      0   52.0000      NaN  
486      0   90.0000      C93  
3        0   53.1000     C123  
1        0   71.2833      C85  
577      0   55.9000      E44  
52       0   76.7292      D33  
765      0   77.9583      D11  
591      0   78.2667      D20  
513      0   59.4000      NaN  
366      0   75.2500      D37  
31       0  146.5208      B78  
334      0  133.6500      NaN  
375      0   82.1708      NaN  
457      0   51.8625      D21  
669      0   52.0000     C126  
849      0   89.1042      C92

In [200]:
change_db_cleaned.loc[change_db_cleaned['PassengerId']==376,'Age']=22

In [202]:
change_db_cleaned.loc[change_db_cleaned['PassengerId']==670,'Age']=42

In [205]:
change_db_cleaned.loc[change_db_cleaned['PassengerId']==335,'Age']=44

In [207]:
change_db_cleaned.loc[change_db_cleaned['PassengerId']==850,'Age']=43

In [208]:
change_db_cleaned.loc[(change_db_cleaned['Pclass']==1)&(change_db_cleaned['Title']=='Mrs')&(change_db_cleaned['SibSp']==1)&(change_db_cleaned['Parch']==0),['Pclass','Name','Age','SibSp','Parch','Fare','Cabin']].sort_values('Age')

Pclass                                               Name   Age  SibSp  \
307       1  Penasco y Castellana, Mrs. Victor de Satode (M...  17.0      1   
781       1          Dick, Mrs. Albert Adrian (Vera Gillespie)  17.0      1   
700       1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  18.0      1   
291       1            Bishop, Mrs. Dickinson H (Helen Walton)  19.0      1   
151       1                  Pears, Mrs. Thomas (Edith Wearne)  22.0      1   
375       1              Meyer, Mrs. Edgar Joseph (Leila Saks)  22.0      1   
809       1     Chambers, Mrs. Norman Campbell (Bertha Griggs)  33.0      1   
230       1       Harris, Mrs. Henry Birkhardt (Irene Wallach)  35.0      1   
383       1  Holverson, Mrs. Alexander Oskar (Mary Aline To...  35.0      1   
486       1    Hoyt, Mrs. Frederick Maxfield (Jane Anne Forby)  35.0      1   
3         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0      1   
1         1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0      1   
577       1          Silvey, Mrs. William Baird (Alice Munger)  39.0      1   
669       1  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)  42.0      1   
849       1       Goldenberg, Mrs. Samuel L (Edwiga Grabowska)  43.0      1   
334       1  Frauenthal, Mrs. Henry William (Clara Heinshei...  44.0      1   
52        1           Harper, Mrs. Henry Sleeper (Myna Haxtun)  49.0      1   
765       1               Hogeboom, Mrs. John C (Anna Andrews)  51.0      1   
591       1    Stephenson, Mrs. Walter Bertram (Martha Eustis)  52.0      1   
513       1     Rothschild, Mrs. Martin (Elizabeth L. Barrett)  54.0      1   
366       1   Warren, Mrs. Frank Manley (Anna Sophia Atkinson)  60.0      1   
31        1     Spencer, Mrs. William Augustus (Marie Eugenie)   NaN      1   
457       1                  Kenyon, Mrs. Frederick R (Marion)   NaN      1   

     Parch      Fare    Cabin  
307      0  108.9000      C65  
781      0   57.0000      B20  
700      0  227.5250  C62 C64  
291      0   91.0792      B49  
151      0   66.6000       C2  
375      0   82.1708      NaN  
809      0   53.1000       E8  
230      0   83.4750      C83  
383      0   52.0000      NaN  
486      0   90.0000      C93  
3        0   53.1000     C123  
1        0   71.2833      C85  
577      0   55.9000      E44  
669      0   52.0000     C126  
849      0   89.1042      C92  
334      0  133.6500      NaN  
52       0   76.7292      D33  
765      0   77.9583      D11  
591      0   78.2667      D20  
513      0   59.4000      NaN  
366      0   75.2500      D37  
31       0  146.5208      B78  
457      0   51.8625      D21

In [209]:
change_db_cleaned.loc[(change_db_cleaned['Pclass']==1)&(change_db_cleaned['Title']=='Mrs')&(change_db_cleaned['SibSp']==1)&(change_db_cleaned['Parch']==0),['Pclass','Name','Age','SibSp','Parch','Fare','Cabin']].describe()

Pclass        Age  SibSp  Parch        Fare
count    23.0  21.000000   23.0   23.0   23.000000
mean      1.0  36.190476    1.0    0.0   84.038043
std       0.0  13.136282    0.0    0.0   40.162538
min       1.0  17.000000    1.0    0.0   51.862500
25%       1.0  22.000000    1.0    0.0   56.450000
50%       1.0  35.000000    1.0    0.0   76.729200
75%       1.0  44.000000    1.0    0.0   89.552100
max       1.0  60.000000    1.0    0.0  227.525000

In [212]:
change_db_cleaned.loc[(change_db_cleaned['Pclass']==1)&(change_db_cleaned['Title']=='Mrs')&(change_db_cleaned['SibSp']==1)&(change_db_cleaned['Parch']==0)&(change_db_cleaned['Age'].isnull()),['Age']]=36

In [ ]:
change_db_cleaned.loc[(change_db_cleaned['Pclass']==1)&(change_db_cleaned['Title']=='Mrs')&(change_db_cleaned['SibSp']==1)&(change_db_cleaned['Parch']==0),['Pclass','Name','Age','SibSp','Parch','Fare','Cabin']].sort_values('Age')

Pclass                                               Name   Age  SibSp  \
307       1  Penasco y Castellana, Mrs. Victor de Satode (M...  17.0      1   
781       1          Dick, Mrs. Albert Adrian (Vera Gillespie)  17.0      1   
700       1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  18.0      1   
291       1            Bishop, Mrs. Dickinson H (Helen Walton)  19.0      1   
151       1                  Pears, Mrs. Thomas (Edith Wearne)  22.0      1   
375       1              Meyer, Mrs. Edgar Joseph (Leila Saks)  22.0      1   
809       1     Chambers, Mrs. Norman Campbell (Bertha Griggs)  33.0      1   
230       1       Harris, Mrs. Henry Birkhardt (Irene Wallach)  35.0      1   
383       1  Holverson, Mrs. Alexander Oskar (Mary Aline To...  35.0      1   
486       1    Hoyt, Mrs. Frederick Maxfield (Jane Anne Forby)  35.0      1   
3         1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0      1   
31        1     Spencer, Mrs. William Augustus (Marie Eugenie)  36.0      1   
457       1                  Kenyon, Mrs. Frederick R (Marion)  36.0      1   
1         1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0      1   
577       1          Silvey, Mrs. William Baird (Alice Munger)  39.0      1   
669       1  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)  42.0      1   
849       1       Goldenberg, Mrs. Samuel L (Edwiga Grabowska)  43.0      1   
334       1  Frauenthal, Mrs. Henry William (Clara Heinshei...  44.0      1   
52        1           Harper, Mrs. Henry Sleeper (Myna Haxtun)  49.0      1   
765       1               Hogeboom, Mrs. John C (Anna Andrews)  51.0      1   
591       1    Stephenson, Mrs. Walter Bertram (Martha Eustis)  52.0      1   
513       1     Rothschild, Mrs. Martin (Elizabeth L. Barrett)  54.0      1   
366       1   Warren, Mrs. Frank Manley (Anna Sophia Atkinson)  60.0      1   

     Parch      Fare    Cabin  
307      0  108.9000      C65  
781      0   57.0000      B20  
700      0  227.5250  C62 C64  
291      0   91.0792      B49  
151      0   66.6000       C2  
375      0   82.1708      NaN  
809      0   53.1000       E8  
230      0   83.4750      C83  
383      0   52.0000      NaN  
486      0   90.0000      C93  
3        0   53.1000     C123  
31       0  146.5208      B78  
457      0   51.8625      D21  
1        0   71.2833      C85  
577      0   55.9000      E44  
669      0   52.0000     C126  
849      0   89.1042      C92  
334      0  133.6500      NaN  
52       0   76.7292      D33  
765      0   77.9583      D11  
591      0   78.2667      D20  
513      0   59.4000      NaN  
366      0   75.2500      D37

In [215]:
change_db2.loc[(change_db2['Pclass']==1)&(change_db2['Title']=='Mrs')&(change_db2['SibSp']==1)&(change_db2['Parch']==0),['Pclass','Name','Age','SibSp','Parch','Fare']].describe()

Pclass        Age  SibSp  Parch        Fare
count    23.0  17.000000   23.0   23.0   23.000000
mean      1.0  35.823529    1.0    0.0   84.038043
std       0.0  13.933930    0.0    0.0   40.162538
min       1.0  17.000000    1.0    0.0   51.862500
25%       1.0  22.000000    1.0    0.0   56.450000
50%       1.0  35.000000    1.0    0.0   76.729200
75%       1.0  49.000000    1.0    0.0   89.552100
max       1.0  60.000000    1.0    0.0  227.525000

In [214]:
change_db_cleaned.loc[(change_db_cleaned['Pclass']==1)&(change_db_cleaned['Title']=='Mrs')&(change_db_cleaned['SibSp']==1)&(change_db_cleaned['Parch']==0),['Pclass','Name','Age','SibSp','Parch','Fare','Cabin']].describe()

Pclass        Age  SibSp  Parch        Fare
count    23.0  23.000000   23.0   23.0   23.000000
mean      1.0  36.173913    1.0    0.0   84.038043
std       0.0  12.525074    0.0    0.0   40.162538
min       1.0  17.000000    1.0    0.0   51.862500
25%       1.0  27.500000    1.0    0.0   56.450000
50%       1.0  36.000000    1.0    0.0   76.729200
75%       1.0  43.500000    1.0    0.0   89.552100
max       1.0  60.000000    1.0    0.0  227.525000

#### Age Imputation – Mrs (`1st Class`, `1 SibSp`, `0 Parch`)
- Group: **23** `Mrs.` passengers in 1st class with 1 SibSp, 0 Parch

- **6** had missing Age values

Imputation Strategy:
- For 4 accompanied by husbands → **imputed as Husband Age - 6** (based on avg age gap)

- For remaining 2 → **used group’s updated mean age after first 4 imputations**

In [90]:
change_db2.loc[((change_db2['Pclass']==3)&(change_db2['Title']=='Mrs')),'Age'].describe()

count    33.000000
mean     33.515152
std      10.031579
min      15.000000
25%      27.000000
50%      31.000000
75%      40.000000
max      63.000000
Name: Age, dtype: float64

In [91]:
change_db2.loc[((change_db2['Pclass']==3)&(change_db2['Title']=='Mrs')),'Age'].median()

np.float64(31.0)

In [79]:
change_db2.loc[(change_db2['Age'].isnull())&(change_db2['Title']=='Mrs'),['Name','Pclass','SibSp','Parch','Fare']].sort_values('Pclass')

Name  Pclass  SibSp  Parch  \
31      Spencer, Mrs. William Augustus (Marie Eugenie)       1      1      0   
166             Chibnall, Mrs. (Edith Martha Bowerman)       1      0      1   
334  Frauenthal, Mrs. Henry William (Clara Heinshei...       1      1      0   
256                     Thorne, Mrs. Gertrude Maybelle       1      0      0   
457                  Kenyon, Mrs. Frederick R (Marion)       1      1      0   
669  Taylor, Mrs. Elmer Zebley (Juliet Cummins Wright)       1      1      0   
375              Meyer, Mrs. Edgar Joseph (Leila Saks)       1      1      0   
849       Goldenberg, Mrs. Samuel L (Edwiga Grabowska)       1      1      0   
347          Davison, Mrs. Thomas Henry (Mary E Finck)       3      1      0   
186    O'Brien, Mrs. Thomas (Johanna "Hannah" Godfrey)       3      1      0   
19                             Masselmani, Mrs. Fatima       3      0      0   
367                     Moussa, Mrs. (Mantoura Boulos)       3      0      0   
431  Thorneycroft, Mrs. Percival (Florence Kate White)       3      1      0   
415            Meek, Mrs. Thomas (Annie Louise Rowley)       3      0      0   
533             Peter, Mrs. Catherine (Catherine Rizk)       3      0      2   
140                      Boulos, Mrs. Joseph (Sultana)       3      0      2   
578                   Caram, Mrs. Joseph (Maria Elias)       3      1      0   

         Fare  
31   146.5208  
166   55.0000  
334  133.6500  
256   79.2000  
457   51.8625  
669   52.0000  
375   82.1708  
849   89.1042  
347   16.1000  
186   15.5000  
19     7.2250  
367    7.2292  
431   16.1000  
415    8.0500  
533   22.3583  
140   15.2458  
578   14.4583

## 🧹 Final Age Imputation Decisions

- **`Dr.` title (1 null):** Imputed with **40.6**, the mean age of male Dr. passengers.  
- **`Master` title in 3rd class (4 nulls):** Imputed with **5.3**, the mean age of Master passengers in 3rd class.

### 👩‍🦳 Age Imputation for 'Mrs.' Title (1st Class)

- **0 SibSp, 0 Parch**  
  - 1 missing age imputed as **57**  
  - Based on group mean (51.3), max age (62), and similar fare (79.2), suggesting an older, wealthy widow.

- **0 SibSp, 1 Parch**  
  - 1 missing age imputed as **50**  
  - Based on group mean (50.2) and presence of a daughter aged 22, estimating maternal age.

- **1 SibSp, 0 Parch** (23 passengers, 6 nulls)  
  - 4 passengers accompanied by husbands → imputed as **Husband Age - 6**  
    - Resulting ages: `22, 42, 44, 43`  
  - 2 remaining passengers (likely with siblings) → imputed with **group’s updated mean age** = **36**

